scripts/validate_min_versions_in_sync.py

```diff
@@ -41,6 +46,84 @@
+            yaml_deps = res
+            for dep_line in yaml_deps:
+                replace_text = operator = yaml_left = ""
+                search_text = str(dep_line)
+                if str(dep_line) in exclusion_list:
+                    continue
+                if ">=" in dep_line:
+                    operator = ">="
+                elif "==" in dep_line:
+                    operator = "=="
+                elif "=" in dep_line:
+                    operator = "="
+                elif "<" in dep_line:
+                    operator = "<"
+                elif ">" in dep_line:
+                    operator = ">"
+                else:
+                    operator = ""
+                if operator == "":
+                    yaml_package, yaml_version = str(dep_line).strip(), None
+                    yaml_left = yaml_package
+                else:
+                    yaml_package, yaml_version = str(dep_line).strip().split(operator)
+                    if operator == "<" or operator == ">":
+                        if yaml_package in toml_dependencies:
+                            if version.parse(yaml_version) <= version.parse(
+                                toml_dependencies[yaml_package]
+                            ):
+                                yaml_left = yaml_package
+                            else:
+                                yaml_left = str(dep_line) + ", "
+                    else:
+                        yaml_left = yaml_package + operator
+                if yaml_package in toml_dependencies:
+                    if ">" in yaml_left or "<" in yaml_left:
+                        if "," in yaml_left:
+                            # ex: "numpy<1.24.0," + ">=" + "1.2"
+                            replace_text = (
+                                yaml_left + ">=" + toml_dependencies[yaml_package]
+                            )
+                        else:
+                            replace_text = yaml_left + toml_dependencies[yaml_package]
+                    # update yaml package version to TOML min version
+                    elif yaml_version is not None:
+                        if version.parse(
+                            toml_dependencies[yaml_package]
+                        ) > version.parse(yaml_version):
+                            # ex: "hypothesis>=" + "6.34.2"
+                            replace_text = yaml_left + toml_dependencies[yaml_package]
+                        elif version.parse(
+                            toml_dependencies[yaml_package]
+                        ) == version.parse(yaml_version):
+                            replace_text = dep_line
+                    else:
+                        # ex: "hypothesis + ">=" + 6.34.2"
+                        replace_text = (
+                            yaml_package + ">=" + toml_dependencies[yaml_package]
+                        )
+                    data = data.replace(search_text, replace_text)
```

Comment:
Sounds good! I've separated this portion of code into its own function. I'm writing unit tests now.

Reply:
```
Unit Test Design

Properties:
- dummy_yaml_answer_files =     [o1.yaml, o2.yaml, o3.yaml, o4.yaml]
- dummy_yaml_unmodified_files = [i1.yaml, i2.yaml, i3.yaml, i4.yaml]
- dummy_toml_file =             path/to/file/dummy_min_dependencies.toml

- Test case 1  
       - result_yaml_file_str = test_update_yaml_file_version(dummy_toml_file, dummy_yaml_unmodified_files[0])
       - expected_yaml_file_str = dummy_yaml_answer_files[0] 
       - assert result_yaml_file_str == expected_yaml_file_str                
- Test case 2
       - result_yaml_file_str = test_update_yaml_file_version(dummy_toml_file, dummy_yaml_unmodified_files[1])
       - expected_yaml_file_str = dummy_yaml_answer_files[1]
       - assert result_yaml_file_str == expected_yaml_file_str               
- Test case 3
       - result_yaml_file_str = test_update_yaml_file_version(dummy_toml_file, dummy_yaml_unmodified_files[2])
       - expected_yaml_file_str = dummy_yaml_answer_files[2]           
       - assert result_yaml_file_str == expected_yaml_file_str       
```        

I was thinking in the code base, there would be a few sets of dummy YAML unmodified/answer files + 1 dummy TOML file. These dummy files would be solely used to ensure the function that is being tested is working as expected. 

Am I thinking about this correctly? Some concerns I have is that it may not be an accurate form of testing since the `dummy_yaml_answer_files` are manually created to produce those "expected results". Also, I'm not sure how effective the test cases would be if I am only testing for equality between 2 files. Would it be unhelpful as a test case since each output file contains such a large dataset? For example, it would not be able to tell which specific dependency version of the file is wrong. It would only inform that the 2 files are not equal to one another.

@MarcoGorelli Could you please provide insight into the matter? I will appreciate it!

Reply:
I wrote a new unit test file at `scripts/tests/test_validate_min_versions_in_sync.py`.

It comes with 3 additional files under scripts/tests/:
```
- dummy_toml_file.toml
- dummy_yaml_expected_file_1.yaml
- dummy_yaml_unmodified_file_1.yaml
```

The .toml file can be adjusted to change the baselines for the minimum dependencies. The `dummy_yaml_unmodified_file_1.yaml` can be adjusted to test different edge cases, however, the `dummy_yaml_expected_file_1.yaml` should be changed accordingly to reflect the proper results for the previously mentioned file, since the two should be in sync as a testing set.

I was thinking of creating more sets of expected/unmodified YAML files so that each set can test its own edge case like duplicates, same version, no version, range, etc. That way, each unit test can be smaller and more meaningful.

Currently, I think the dataset is too large to provide meaningful test assertion results.

@MarcoGorelli What are your thoughts? Feedback appreciated!

Reply:
nice! this isn't user-facing, so arguably doesn't need too extensive testing. but yes, smaller test files would make it easier to reason about what the script is doing